File: sdk/ruby/lib/velocity_sdk/workflow_options.rb
# frozen_string_literal: true

module VelocitySdk
  # Builder for workflow execution options.
  #
  # Provides a fluent interface for configuring workflow parameters such as
  # namespace, task queue, timeouts, retry policy, and execution steps.
  #
  # @example
  #   options = VelocitySdk::WorkflowOptions.new
  #     .namespace('production')
  #     .task_queue('high-priority')
  #     .total_steps(10)
  #     .execution_timeout_ms(30_000)
  #     .retry_policy(max_attempts: 3)
  class WorkflowOptions
    # @return [String] Namespace to run in.
    attr_reader :namespace

    # @return [String] Task queue for worker dispatch.
    attr_reader :task_queue

    # @return [Integer] Number of execution steps.
    attr_reader :total_steps

    # @return [Integer] Execution timeout in milliseconds.
    attr_reader :execution_timeout_ms

    # @return [Hash] Retry policy configuration.
    attr_reader :retry_policy

    # @return [String] Explicit workflow ID (empty = server-assigned).
    attr_reader :workflow_id

    # @return [Hash] Search attributes for visibility queries.
    attr_reader :search_attributes

    # @return [String, nil] Memo attached to the workflow.
    attr_reader :memo

    # Create a new options builder with defaults.
    def initialize
      @namespace = 'default'
      @task_queue = 'default'
      @total_steps = 1
      @execution_timeout_ms = 60_000
      @retry_policy = {}
      @workflow_id = ''
      @search_attributes = {}
      @memo = nil
    end

    # Create options with all defaults.
    # @return [WorkflowOptions]
    def self.defaults
      new
    end

    # Set the namespace.
    # @param value [String]
    # @return [self]
    def namespace(value)
      @namespace = value
      self
    end

    # Set the task queue.
    # @param value [String]
    # @return [self]
    def task_queue(value)
      @task_queue = value
      self
    end

    # Set the total number of execution steps.
    # @param value [Integer]
    # @return [self]
    def total_steps(value)
      @total_steps = [1, value].max
      self
    end

    # Set the execution timeout in milliseconds.
    # @param value [Integer]
    # @return [self]
    def execution_timeout_ms(value)
      @execution_timeout_ms = [0, value].max
      self
    end

    # Set the retry policy.
    # @param value [Hash]
    # @return [self]
    def retry_policy(value)
      @retry_policy = value
      self
    end

    # Set an explicit workflow ID.
    # @param value [String]
    # @return [self]
    def workflow_id(value)
      @workflow_id = value
      self
    end

    # Set search attributes.
    # @param value [Hash]
    # @return [self]
    def search_attributes(value)
      @search_attributes = value
      self
    end

    # Set a memo.
    # @param value [String]
    # @return [self]
    def memo(value)
      @memo = value
      self
    end

    # Convert to a Hash for serialisation.
    # @return [Hash]
    def to_h
      {
        namespace: @namespace,
        task_queue: @task_queue,
        total_steps: @total_steps,
        execution_timeout_ms: @execution_timeout_ms,
        retry_policy: @retry_policy,
        workflow_id: @workflow_id,
        search_attributes: @search_attributes,
        memo: @memo,
      }
    end

    # @return [String]
    def inspect
      "#<VelocitySdk::WorkflowOptions ns=#{@namespace} tq=#{@task_queue} steps=#{@total_steps}>"
    end
  end
end

File: sdk/php/src/Workflow/WorkflowOptions.php
<?php

declare(strict_types=1);

namespace Velocity\SDK\Workflow;

/**
 * Builder for workflow execution options.
 *
 * Provides a fluent interface for configuring workflow parameters such as
 * namespace, task queue, timeouts, retry policy, and execution steps.
 *
 * Usage:
 *     $options = WorkflowOptions::new()
 *         ->withNamespace('production')
 *         ->withTaskQueue('high-priority')
 *         ->withTotalSteps(10)
 *         ->withExecutionTimeoutMs(30_000)
 *         ->withRetryPolicy(['max_attempts' => 3]);
 */
class WorkflowOptions
{
    private string $namespace = 'default';
    private string $taskQueue = 'default';
    private int $totalSteps = 1;
    private int $executionTimeoutMs = 60_000;
    private array $retryPolicy = [];
    private string $workflowId = '';
    private array $searchAttributes = [];
    private ?string $memo = null;

    /** Use the named constructor {@see defaults()} or {@see new()}. */
    private function __construct()
    {
    }

    /** Create a new options builder. */
    public static function new(): self
    {
        return new self();
    }

    /** Create options with all defaults. */
    public static function defaults(): self
    {
        return new self();
    }

    /** Set the namespace. */
    public function withNamespace(string $namespace): self
    {
        $this->namespace = $namespace;
        return $this;
    }

    /** Set the task queue. */
    public function withTaskQueue(string $taskQueue): self
    {
        $this->taskQueue = $taskQueue;
        return $this;
    }

    /** Set the total number of execution steps. */
    public function withTotalSteps(int $totalSteps): self
    {
        $this->totalSteps = max(1, $totalSteps);
        return $this;
    }

    /** Set the execution timeout in milliseconds. */
    public function withExecutionTimeoutMs(int $timeoutMs): self
    {
        $this->executionTimeoutMs = max(0, $timeoutMs);
        return $this;
    }

    /** Set the retry policy as an associative array. */
    public function withRetryPolicy(array $policy): self
    {
        $this->retryPolicy = $policy;
        return $this;
    }

    /** Set an explicit workflow ID (server assigns one if empty). */
    public function withWorkflowId(string $workflowId): self
    {
        $this->workflowId = $workflowId;
        return $this;
    }

    /** Set search attributes for visibility queries. */
    public function withSearchAttributes(array $attributes): self
    {
        $this->searchAttributes = $attributes;
        return $this;
    }

    /** Set a memo attached to the workflow. */
    public function withMemo(string $memo): self
    {
        $this->memo = $memo;
        return $this;
    }

    // ─── Accessors ────────────────────────────────────────────────────────

    public function getNamespace(): string { return $this->namespace; }
    public function getTaskQueue(): string { return $this->taskQueue; }
    public function getTotalSteps(): int { return $this->totalSteps; }
    public function getExecutionTimeoutMs(): int { return $this->executionTimeoutMs; }
    public function getRetryPolicy(): array { return $this->retryPolicy; }
    public function getWorkflowId(): string { return $this->workflowId; }
    public function getSearchAttributes(): array { return $this->searchAttributes; }
    public function getMemo(): ?string { return $this->memo; }

    /**
     * Convert to an associative array for serialisation.
     *
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return [
            'namespace' => $this->namespace,
            'task_queue' => $this->taskQueue,
            'total_steps' => $this->totalSteps,
            'execution_timeout_ms' => $this->executionTimeoutMs,
            'retry_policy' => $this->retryPolicy,
            'workflow_id' => $this->workflowId,
            'search_attributes' => $this->searchAttributes,
            'memo' => $this->memo,
        ];
    }
}
